lib/graph.rb

```diff
@@ -172,20 +181,119 @@ def reverse!
         @vertices = v
         @edge_labels = l
       end
+
       ##
       # Returns array of vertices
       def vertices
         @vertices
       end
 
+      ##
+      # Finds Strongly Connected Components (SCC) in graph. SCC is a subgraph where every vertex is reachable from every other vertex.
+      # @return Array of SCC. Each component is represented as Array of vertices.
+      # @author vaimon
+      def find_strongly_connected_components
+        # Vertices we have already visited.
+        visited = Hash.new(false)
+        # Timestamps got during depth-first search.
+        order = []
+        # Resulting array of SCC.
+        res = []
+
+        # Step 1: Launch DFS to get order marks
+        @vertices.keys.each do |key|
+          unless visited[key]
+            tuple = scc_dfs_first key, visited, order
+            visited, order = tuple
+          end
+        end
+
+        # Step 2: Transpose adjacency list
+        transposed = transpose_adjacency_list
+
+        # Step 3: Launch second DFS in reverse order of timestamps from Step 1 to build components.
+        visited = Hash.new(false)
+        order.reverse.each do |v|
+          unless visited[v]
+            component = []
+            tuple = scc_dfs_second v, component, visited, transposed
```

Comment:
is `tuple` really needed?

We can just 
`visited, component = scc_dfs_second v, component, visited, transposed`